data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java

```diff
@@ -137,41 +139,44 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
-                            new OutputCodecContext(),
-                            (inputBuffer, response) -> {
-                                Duration latency = inputBuffer.stopLatencyWatch();
-                                lambdaLatencyMetric.record(latency.toMillis(), TimeUnit.MILLISECONDS);
-                                numberOfRecordsSuccessCounter.increment(inputBuffer.getEventCount());
-                                numberOfRequestsSuccessCounter.increment();
-                                return convertLambdaResponseToEvent(inputBuffer, response);
-                            },
-                            (inputBuffer) -> {
-                                Duration latency = inputBuffer.stopLatencyWatch();
-                                lambdaLatencyMetric.record(latency.toMillis(), TimeUnit.MILLISECONDS);
-                                numberOfRecordsFailedCounter.increment(inputBuffer.getEventCount());
-                                numberOfRequestsFailedCounter.increment();
-                                return addFailureTags(inputBuffer.getRecords());
-                            })
-
-            );
-        } catch (Exception e) {
-            LOG.info("Exception in doExecute");
-            numberOfRecordsFailedCounter.increment(recordsToLambda.size());
-            resultRecords.addAll(addFailureTags(recordsToLambda));
+
+        Map<Buffer, CompletableFuture<InvokeResponse>> bufferToFutureMap = LambdaCommonHandler.sendRecords(
+            recordsToLambda, lambdaProcessorConfig, lambdaAsyncClient,
+            new OutputCodecContext());
+
+        for (Map.Entry<Buffer, CompletableFuture<InvokeResponse>> entry : bufferToFutureMap.entrySet()) {
+            CompletableFuture<InvokeResponse> future = entry.getValue();
+            Buffer inputBuffer = entry.getKey();
+            try {
+                InvokeResponse response = future.join();
+                Duration latency = inputBuffer.stopLatencyWatch();
+                lambdaLatencyMetric.record(latency.toMillis(), TimeUnit.MILLISECONDS);
+                if (isSuccess(response)) {
+                    numberOfRecordsSuccessCounter.increment(inputBuffer.getEventCount());
+                    numberOfRequestsSuccessCounter.increment();
+                    resultRecords.addAll(convertLambdaResponseToEvent(inputBuffer, response));
+                } else {
+                    LOG.error("Lambda invoke failed with error {} ", response.statusCode());
+                    resultRecords.addAll(addFailureTags(inputBuffer.getRecords()));
+                }
+            } catch (Exception e) {
+                LOG.error("Exception from Lambda invocation ", e);
+                numberOfRecordsFailedCounter.increment(inputBuffer.getEventCount());
+                numberOfRequestsFailedCounter.increment();
+                resultRecords.addAll(addFailureTags(inputBuffer.getRecords()));
```

Comment:
optionally, do the same here. Though it really doesn't much difference in this case.

Reply:
As we discussed with Krishan, he will address this change in his next PR